python/src/fast.cpp

```diff
@@ -183,4 +188,102 @@ void init_fast(nb::module_& parent_module) {
         Returns:
           array: The quantized version of ``w``
       )pbdoc");
+
+  m.def(
+      "metal_kernel",
+      [](std::string name,
+         std::string source,
+         std::map<std::string, nb::handle> inputs_,
+         std::map<std::string, std::vector<int>> output_shapes,
+         std::map<std::string, Dtype> output_dtypes,
+         std::tuple<int, int, int> grid,
+         std::tuple<int, int, int> threadgroup,
+         std::optional<std::map<std::string, nb::handle>> template_args_,
+         bool ensure_row_contiguous,
+         bool verbose,
+         StreamOrDevice s) -> std::map<std::string, array> {
+        std::map<std::string, fast::TemplateArg> template_args;
+        if (template_args_) {
+          for (auto [name, value] : template_args_.value()) {
+            // Handle bool, int and dtype template args
+            if (PyBool_Check(value.ptr())) {
+              bool bool_val = nb::cast<bool>(value);
+              template_args.insert({name, bool_val});
+            } else if (PyLong_Check(value.ptr())) {
+              int int_val = nb::cast<int>(value);
+              template_args.insert({name, int_val});
+            } else if (
+                nb::type_check(value.type()) &&
+                nb::type_info(value.type()) == typeid(Dtype)) {
+              Dtype dtype = nb::cast<Dtype>(value);
+              template_args.insert({name, dtype});
+            } else {
+              throw std::invalid_argument(
+                  "Invalid template argument. Must be `mlx.core.Dtype`, `int` or `bool`.");
+            }
+          }
+        }
+        std::map<std::string, array> inputs;
+        for (auto [name, in] : inputs_) {
+          if (name != "stream") {
+            auto value = nb::cast<ArrayInitType>(in);
+            auto arr = create_array(value, std::nullopt);
+            inputs.insert({name, arr});
```

Comment:
If you switch to `ScalarOrArray` that will simplify to
```suggestion
            inputs.insert({name, to_array(arr)});
```